drivers/afe/ad413x/ad413x.c

```diff
@@ -0,0 +1,816 @@
+/******************************************************************************/
+#include <stdio.h>
+#include <stdlib.h>
+#include "ad413x.h"
+#include "no-os/error.h"
+#include "no-os/irq.h"
+#include "no-os/print_log.h"
+#include "no-os/delay.h"
+
+/******************************************************************************/
+/************************** Functions Implementation **************************/
+/******************************************************************************/
+/***************************************************************************//**
+ * Compute CRC8 checksum.
+ *
+ * @param data      - The data buffer.
+ * @param data_size - The size of the data buffer.
+ *
+ * @return CRC8 checksum.
+*******************************************************************************/
+static uint8_t ad413x_compute_crc8(uint8_t *data, uint8_t data_size)
+{
+	uint8_t i;
+	uint8_t crc = 0;
+
+	while (data_size) {
+		for (i = 0x80; i != 0; i >>= 1) {
+			if (((crc & 0x80) != 0) != ((*data & i) != 0)) {
+				crc <<= 1;
+				crc ^= AD413X_CRC8_POLY;
+			} else
+				crc <<= 1;
+		}
+		data++;
+		data_size--;
+	}
+	return crc;
+}
+
+/***************************************************************************//**
+ * SPI DATA register continuous read from device.
+ *
+ * @param dev      - The device structure.
+ * @param reg_data - The register data.
+ *
+ * @return SUCCESS in case of success, negative error code otherwise.
+*******************************************************************************/
+static int32_t ad413x_spi_data_cont_read(struct ad413x_dev *dev,
+		uint32_t *reg_data)
+{
+	uint8_t ret;
+	uint8_t buf[3];
+	uint32_t data;
+
+	memset(buf, 0xAA, 3); // dummy data bytes
+
+	ret = spi_write_and_read(dev->spi_dev, buf, 3);
+	if (ret)
+		return ret;
+
+	data = (buf[0] << 16) | (buf[1] << 8) | buf[2];
+
+	*reg_data = data;
+
+	return SUCCESS;
+}
+
+/***************************************************************************//**
+ * IRQ handler for single ADC read.
+*******************************************************************************/
+static void irq_adc_single_read(struct ad413x_callback_ctx *ctx)
+{
+	struct ad413x_dev *dev = ctx->dev;
+	if(ctx->buffer_size > 0) {
+		if(ad413x_spi_int_reg_read(ctx->dev, AD413X_REG_DATA, ctx->buffer))
+			pr_err("DATA reg could not be read \n");
+		ctx->buffer_size--;
+		ctx->buffer++;
+		if(irq_enable(dev->irq_desc, dev->rdy_pin))
+			pr_err("IRQ_enable error \n");
+	} else if(irq_disable(dev->irq_desc, dev->rdy_pin)) {
+		pr_err("IRQ_disable error \n");
+	}
+}
+
+/***************************************************************************//**
+ * IRQ handler for continuously ADC read.
+*******************************************************************************/
+static void irq_adc_cont_read(struct ad413x_callback_ctx *ctx)
+{
+	struct ad413x_dev *dev = ctx->dev;
+	if(ctx->buffer_size > 0) {
+		if(ad413x_spi_data_cont_read(ctx->dev, ctx->buffer))
```

Comment:
Since the code here is almost identical with `irq_adc_single_read` maybe we could include a flag in the ctx handler and call different functions here instead of almost duplicating the code. Do you think this makes sense or would functions would be generally different usually?